pkg/attribute/tracker_test.go

```diff
@@ -86,3 +86,232 @@ func BenchmarkTracker(b *testing.B) {
 		}
 	}
 }
+
+func TestTracker_StartRequest(t *testing.T) {
+	t9 := time.Date(2001, 1, 1, 1, 1, 1, 9, time.UTC)
+	t10 := time.Date(2001, 1, 1, 1, 1, 1, 10, time.UTC)
+	ts9, _ := ptypes.TimestampProto(t9)
+	ts10, _ := ptypes.TimestampProto(t10)
+	d := time.Duration(42) * time.Second
+	ds := ptypes.DurationProto(d)
+	sm := &mixerpb.StringMap{Map: map[int32]string{14: "14"}}
+
+	attr1 := mixerpb.Attributes{
+		Dictionary: dictionary{1: "N1", 2: "N2", 3: "N3", 4: "N4", 5: "N5", 6: "N6", 7: "N7", 8: "N8",
+			9: "N9", 10: "N10", 11: "N11", 12: "N12", 13: "N13", 14: "N14"},
+		StringAttributes:    map[int32]string{1: "1", 2: "2"},
+		Int64Attributes:     map[int32]int64{3: 3, 4: 4},
+		DoubleAttributes:    map[int32]float64{5: 5.0, 6: 6.0},
+		BoolAttributes:      map[int32]bool{7: true, 8: false},
+		TimestampAttributes: map[int32]*ptypes.Timestamp{9: ts9, 10: ts10},
+		DurationAttributes:  map[int32]*ptypes.Duration{11: ds},
+		BytesAttributes:     map[int32][]uint8{12: {12}, 13: {13}},
+		StringMapAttributes: map[int32]*mixerpb.StringMap{14: sm},
+	}
+
+	attr2 := mixerpb.Attributes{
+		Dictionary: dictionary{1: "X1", 2: "X2", 3: "X3", 4: "X4", 5: "X5", 6: "X6", 7: "X7", 8: "X8",
+			9: "X9", 10: "X10", 11: "X11", 12: "X12", 13: "X13", 14: "X14"},
+		StringAttributes:    map[int32]string{1: "1", 2: "2"},
+		Int64Attributes:     map[int32]int64{3: 3, 4: 4},
+		DoubleAttributes:    map[int32]float64{5: 5.0, 6: 6.0},
+		BoolAttributes:      map[int32]bool{7: true, 8: false},
+		TimestampAttributes: map[int32]*ptypes.Timestamp{9: ts9, 10: ts10},
+		DurationAttributes:  map[int32]*ptypes.Duration{11: ds},
+		BytesAttributes:     map[int32][]uint8{12: {12}, 13: {13}},
+		StringMapAttributes: map[int32]*mixerpb.StringMap{14: sm, 15: sm},
+	}
+
+	tracker := NewManager().NewTracker().(*tracker)
+	_, err := tracker.StartRequest(&attr1)
+	if err != nil {
+		t.Errorf("Expecting success, got %v", err)
+	}
+
+	oldDict := tracker.currentDictionary
+	oldBag := copyBag(tracker.contexts[0])
+
+	tracker.StartRequest(&attr2)
+	_, err = tracker.StartRequest(&attr2)
```

Comment:
It's calling StartRequest(attr1) which works, followed by StartRequest(attr2) which fails because of a bad attribute index. The failure is indeed discovered within the checkPrecondition method.